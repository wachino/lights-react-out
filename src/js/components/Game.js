import React from 'react';
import PropTypes from 'prop-types';
import 'css/Game.css';
import Board from 'components/Board';

export default class Game extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      lightStates: Array(props.boardSize).fill().map(
        () => Array(props.boardSize).fill(false)
      )
    };
    this.handleClick = this.handleClick.bind(this);
  }

  render () {
    return (
      <div className='game'>
        <Board
          lightStates={this.state.lightStates}
          onClick={this.handleClick}
        />
      </div>
    );
  }

  handleClick (row, col) {
    const lightStates = this.state.lightStates.slice();
    if (this.isWinner()) {
      return;
    }

    lightStates[row][col] = !lightStates[row][col];
    col > 0 && (lightStates[row][col - 1] = !lightStates[row][col - 1]);
    row > 0 && (lightStates[row - 1][col] = !lightStates[row - 1][col]);
    row < this.props.boardSize - 1 &&
      (lightStates[row + 1][col] = !lightStates[row + 1][col]);
    col < this.props.boardSize - 1 &&
      (lightStates[row][col + 1] = !lightStates[row][col + 1]);

    this.setState({lightStates});
    if (this.isWinner()) {
      alert('Winneeer!!');
    }
  }

  isWinner () {
    return this.state.lightStates.every(
      row => row.every(cell => cell)
    );
  }
}

Game.propTypes = {
  boardSize: PropTypes.number.isRequired
};
