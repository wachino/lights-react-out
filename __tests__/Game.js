import React from 'react';
import Game from '../src/js/components/Game';
import renderer from 'react-test-renderer';

test('Game with no winner', () => {
  var boardSize = 5;

  const component = renderer.create(
    <Game
      boardSize={boardSize}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  for (let father of tree.children) {
    for (let son of father.children) {
      for (let grandchild of son.children) {
        for (let greatgrandson of grandchild.children) {
          greatgrandson.props.onClick();
        }
      }
    }
  }

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Game winner', () => {
  var boardSize = 2;

  const component = renderer.create(
    <Game
      boardSize={boardSize}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  for (let father of tree.children) {
    for (let son of father.children) {
      for (let grandchild of son.children) {
        for (let greatgrandson of grandchild.children) {
          greatgrandson.props.onClick();
        }
      }
    }
  }

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Game winner plus one click more', () => {
  var boardSize = 2;

  const component = renderer.create(
    <Game
      boardSize={boardSize}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  var last;
  for (let father of tree.children) {
    for (let son of father.children) {
      for (let grandchild of son.children) {
        for (let greatgrandson of grandchild.children) {
          greatgrandson.props.onClick();
          last = greatgrandson.props;
        }
      }
    }
  }

  last.onClick();

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
