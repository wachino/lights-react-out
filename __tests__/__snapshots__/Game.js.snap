// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game winner 1`] = `
<div
  className="game"
>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game winner 2`] = `
<div
  className="game"
>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game winner plus one click more 1`] = `
<div
  className="game"
>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game winner plus one click more 2`] = `
<div
  className="game"
>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game with no winner 1`] = `
<div
  className="game"
>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game with no winner 2`] = `
<div
  className="game"
>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
      <div
        className="cell"
      >
        <div
          className="square isOn"
          onClick={[Function]}
        >
          <div
            className="cellOverlay"
          >
            <div
              className="cdot"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
